### ISSUE TITLE
Prevent dev pods from importing private header files of other pods

### ISSUE THREAD

**Issue by lusnaow** [CocoaPods/CocoaPods#8474]
* [x] I've read and understood the [*CONTRIBUTING* guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report

## What did you do?
 import private header files

## What did you expect to happen?
Let dev pods can not import private header files of other pods

## What happened instead?

Dev pods can import private header files of other pods, no matter using import "xxx" or import <xxx/xxx>

## CocoaPods Environment
```
   CocoaPods : 1.6.0.beta.2
        Ruby : ruby 2.6.0p0 (2018-12-25 revision 66547) [x86_64-darwin18]
    RubyGems : 3.0.1
        Host : Mac OS X 10.14.2 (18C54)
       Xcode : 10.0 (10A255)
```

### Installation Source

```
Executable Path: /usr/local/bin/pod
```

### Plugins

```
cocoapods-deintegrate : 1.0.2
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.3.1
cocoapods-try         : 1.1.0
```

**Comment by paulb777** [comment]
What would the point of this change be? If private headers were no longer accessible, they would be no different from internal headers.

**Comment by amorde** [comment]
@paulb777 this is referring to private headers of _other_ pods - shouldn't private headers be private to that specific pod?

**Comment by amorde** [comment]
oh wait I think I see what your saying - I was confusing `Project` and `Private` headers

**Comment by paulb777** [comment]
Yeah, the naming is bad, but private headers are a good way to expose APIs across pods without making them part of the pod's default module import or part of the pod's public API contract.

**Comment by amorde** [comment]
Yeah I agree. It seems like the solution here is to just use `Project` headers, which you can do by including headers in `source_files` and _not_ including them in either `public_header_files` or `private_header_files`

**Comment by paulb777** [comment]
Yes

**Comment by lusnaow** [comment]
@paulb777 Can you provide more info of "internal headers" to me? Thanks.  I put all  headers in `source_files`, but they can be imported to any other development pod. What I want is that one development pod can not import private header files of other development pods. By the way, all my pods are source files, not framework.

**Comment by amorde** [comment]
hmm okay, that sounds like it's not working as expected. If you click on the header you are importing and look at the `Target Membership` section in Xcode, does it say "Project"?

If so, could you upload a sample app that demonstrates the issue?

**Comment by lusnaow** [comment]
> hmm okay, that sounds like it's not working as expected. If you click on the header you are importing and look at the `Target Membership` section in Xcode, does it say "Project"?
> 
> If so, could you upload a sample app that demonstrates the issue?

@amorde Yes, it say "Project". But no matter I change it  to "Private" or "Public", still can be imported.No error. I will upload a sample later.

**Comment by lusnaow** [comment]
[Sample.zip](https://github.com/CocoaPods/CocoaPods/files/2863464/Sample.zip)
@amorde

**Comment by amorde** [comment]
The sample you uploaded doesn't build (it's missing `PodA` from `../PodA`), but I whipped up something that I think reproduces what you're describing.

Given:

* Externally sourced pod *Foo* containing Project header "SomeHeader.h"
* Development Pod *Bar* containing header "SomeObject.h"

Inside `SomeObject.h` of *Bar*, I can successfully import the Project header "SomeHeader.h" when I shouldn't be able to.

I haven't tested other combinations of development pods vs. external pods but I think I can repro this.

Here's my version of your sample 
[Sample.zip](https://github.com/CocoaPods/CocoaPods/files/2863592/Sample.zip)

**Comment by lusnaow** [comment]
> The sample you uploaded doesn't build (it's missing `PodA` from `../PodA`), but I whipped up something that I think reproduces what you're describing.
> 
> Given:
> 
> * Externally sourced pod _Foo_ containing Project header "SomeHeader.h"
> * Development Pod _Bar_ containing header "SomeObject.h"
> 
> Inside `SomeObject.h` of _Bar_, I can successfully import the Project header "SomeHeader.h" when I shouldn't be able to.
> 
> I haven't tested other combinations of development pods vs. external pods but I think I can repro this.
> 
> Here's my version of your sample
> [Sample.zip](https://github.com/CocoaPods/CocoaPods/files/2863604/Sample.zip)
> 
> Edit: Updated sample

Yes, your sample code of `TestProjectHeaderImport.h` is exactly what I want to say.

**Comment by amorde** [comment]
Great, thanks for verifying that. I'm not too sure how this has worked in the past (and I also noticed you are using static libraries, which I haven't personally looked at in a while) so I'm hoping @paulb777 or @dnkoutso could look into this further using my sample

**Comment by paulb777** [comment]
I haven't fully confirmed, but I suspect that if there's an issue may stem from putting the headers in the same directory.  

The visibility works appropriately in the [FirebaseCore](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseCore.podspec#L28) pod by putting the public and private headers in their own directory.  The rest of the headers are not visible.

Also, if `use_frameworks!` is specified, HeaderOne.h is not found since PodB is not a dependency of PodA.

CocoaPods can only do so much about for namespacing if frameworks are disabled.

**Comment by alexlee002** [comment]
@dnkoutso 

> Yes this is a hard issue to fix as it requires breaking the header search paths that has been since version 1.x.

It's so confuse that why the public header files in `static_library` pods not put in the xcode's  public header files' build phase?  And why don't set the header search path to the built product dir, instead of linking it to the `Pods/Headers/Public`?

**Comment by alexlee002** [comment]
And, is there any issue If we add the public header files of the `static library pod` to the `public header build phase`, and then change the search path point to the `built_product_dir/include`?
